### ISSUE TITLE
RMMark, RMPath "vibrate" while the RMMapView is being scrolled

### ISSUE THREAD

**Issue by mouse4d** [mapbox/DEPRECATED-mapbox-ios-sdk#72]
while scrolling the map the RMMarker and RMPath overlay "vibrates", it's like the overlay tries to "catch up" with it's normal position when the map is scrolled. It's just a few pixels, but when zoomed it looks pretty bad. this happens most probably because the -draw() method is called only when the map is moved more than just a pixel.

**Comment by incanus** [comment]
For `RMPath`, see #73. 

For `RMMarker`, I haven't seen this. Do you still see it when not using a path/shape with your markers, and if so, when using `RMShape` instead of `RMPath`?

**Comment by incanus** [comment]
I at least can see this now. Will try to address soon.

**Comment by incanus** [comment]
Assigned to [Tender discussion #259](http://support.mapbox.com/discussions/mapbox-ios-sdk/259-markers-jiggle-when-zoom-is-17).

**Comment by cga-F4** [comment]
Hi, I tried to see where this problem comes from.
With high zoom level, we see that the tile layer moves pixel by pixel while the RMMarker moves only 4 by 4 pixel.
I log some info in the function in RMMapView
- (void)observeValueForKeyPath:(NSString *)aKeyPath ofObject:(id)anObject change:(NSDictionary *)change context:(void *)context
  and we see that we are called each time the tile layer is moved, but the .contentOffset value from the change variable is the same and only changed 4 by 4 pixel. The _mapScroolView.contentOffset follows the same : it does not change pixel by pixel, but 4 by 4 pixel.
  Maybe this is because, with high level zoom, the contentSize and contentOffset are really big value (and there are float), and Apple scrollview does not support accurate move on big big contentOffset/contentSize.

I hope this can help you find a fix for this issue.

**Comment by davidovitch1** [comment]
same problem here, any news ?

**Comment by cga-F4** [comment]
Since last time, i tried using Alpstein Route Me, and we see the same problem with their example MarkerMurder. When you zoom to max, you see the marker vibrating (in this example, you can set the maxZoom to 22, and on maximum zoom, the markers vibrates much more).
I saw the layer with annotations is not in the RMMapScrollView *_mapScrollView. Would it be a bad idea or lots of work to have the RMMapOverlayView *_overlayView and the _tiledLayersSuperview in the same scrollView ?

**Comment by cga-F4** [comment]
I see the problem : when I am on level 19 zoom, and I look to Paris, France, my contentOffset has a value more than 60 000 000. When scrolling, we want to add 1, 2, 3 ... pixels to this value. But due to float precision, 60 000 000 + 1 returns 60 000 000 and 60 000 000 + 3 returns 60 000 004. We have a precision to 4 digit, so we see the annotions move only 4 pixels by 4 pixels. All because contentOffset is a float value (not double). [ By the way, I don't know how the _mapScrollView handles this by itself for the tileLayer which moves perfectly good, maybe the OS computation is in double but we can only access the contentOffset with float precision]
One solution would be to ensure that our contentOffset does not go further than the float precision limit.
From this post : http://stackoverflow.com/questions/1576126/iphone-and-floating-point-math, we can see that if we want a 1 pixel precision, we need to keep contentOffset under approximatively 8 000 000 (log2(8000000) = 22.93 < 23). This explains why we have the problem with big zoom level, and not average zoom level.

**Comment by incanus** [comment]
Interesting, @cga-F4. That's a great find & this makes sense. I will take a look.

**Comment by incanus** [comment]
@trasch, do you have any insight into this? I'm wondering if you ever tried the route of making the overlay layer size with the scroll view's content view, instead of making it a scroll layer with annotation corrections like it is now.

**Comment by trasch** [comment]
I did try it, but I really can't remember what the issue was. Maybe this is not an issue any more, since so many parts of route-me have changed in the meantime. It would certainly make lots of things easier if the overlay would have the size of the content view, maybe we could just try it.

Apart from that I honestly never thought that anyone would use route-me with zoom leves above 18... :)

**Comment by psavich** [comment]
Hello there. I'm experiencing the same problem on zoom levels 17 and above.

Here's a demo: https://dl.dropbox.com/u/586782/mapbox-marker-bug.mp4

The problem is definitely related to float precision of contentOffset – like cga-F4 I injected some logging into the value observer, and here's what I got (each line is a 1px movement):

2012-11-25 19:51:16.403 iGIS[98912:c07] ContentOffset: 162256000.000000 83907544.000000
2012-11-25 19:51:16.420 iGIS[98912:c07] ContentOffset: 162256000.000000 83907544.000000
2012-11-25 19:51:16.436 iGIS[98912:c07] ContentOffset: 162256000.000000 83907544.000000
2012-11-25 19:51:16.453 iGIS[98912:c07] ContentOffset: 162256016.000000 83907544.000000
2012-11-25 19:51:16.470 iGIS[98912:c07] ContentOffset: 162256016.000000 83907544.000000
2012-11-25 19:51:16.487 iGIS[98912:c07] ContentOffset: 162256016.000000 83907536.000000
2012-11-25 19:51:16.503 iGIS[98912:c07] ContentOffset: 162256016.000000 83907536.000000
2012-11-25 19:51:16.520 iGIS[98912:c07] ContentOffset: 162256016.000000 83907536.000000

As you can see, on 19 zoom we have a jump of 16px by x, and 8px by y. Each step of zooming out decreases the "jump distance" by factor of 2.

Unfortunately this issue is critical for our project, cause all RMPaths and RMMarkers vibrate really badly in some important use cases.

I see two solutions:
- unbind annotations from contentOffset and align their position to some more accurate properties.
- workaround by shifting the center of coordinates somewhere near ourselves, to decrease the value of contentOffset and make it more precise (this is possible in our situation because we work with a single city in our project)

What do you guys think about this stuff?

**Comment by zhm** [comment]
+1 to seeing this one fixed. Once this gets fixed I can complete the conversion of 2 apps to this SDK :) This one is just kind of critical because users have maps that are often high resolution satellite imagery.

**Comment by cga-F4** [comment]
Hi ! (and Happy New Year)
Is there some news for this issue ? This is critical for us too.

**Comment by incanus** [comment]
I will be investigating this soon and working with upstream on it. Have a couple ideas here.

**Comment by specialunderwear** [comment]
This blog contains a proper solution to the problem: http://www.mlsite.net/blog/?p=1342 for more context, see https://github.com/Alpstein/route-me/issues/122#issuecomment-11965580

**Comment by incanus** [comment]
I'm not sure this does entirely solve the problem, though I have to investigate further. It says: 

> Note that the content view is set to the size of the entire rendered world; this does not cause out-of-memory problems because the content view does not render any content directly; only those subviews in its visible area are rendered.

Which leads me to believe that we will still have precision problems with higher zoom levels since the effective content view is still dealing in 1:1 pixel sizes with the current method.

**Comment by specialunderwear** [comment]
No, the problem you are seeing, is that on high zoom level you need to scroll 10 pixels for a contentOffset change to be fired, because it is assumed you are viewing a zoomed in bitmap. By resetting the zoomlevel back to 1 you will get back the fine grained scroll events.

**Comment by incanus** [comment]
Ah, interesting.

**Comment by specialunderwear** [comment]
Could you keep this ticket updated with progress? I was planning on working on this myself, but it's better to not waste 2 people's time working on the same thing. Or if you want some help, I'm more then happy to chip in :)

**Comment by incanus** [comment]
Yes, I will. I am working on this combined with improved tile load performance in a hopefully-major (if successful) overhaul of this subsystem.

**Comment by incanus** [comment]
Based on my testing, this is a limitation in `UIScrollView`. I'm currently working on some testing that uses a large (> z17) content view, nothing to do with the KVO calls, and _its own content view_ vibrates in this manner on high zoom. 

I'm working on this issue in combination with performance work from discussions in https://github.com/Alpstein/route-me/issues/35 and will probably end up doing something related to not actually having the scroll view content view be `2^zoom * 256` pixels to a side, but rather "faking out" the system in order to keep the content view within reasonable sizes.

**Comment by specialunderwear** [comment]
Yes, that is true, because again, it is assumed you are looking at a zoomed in bitmap, where it makes no sense to scroll less then 1 (hugely magnified) pixel. Now if you would have tested the kvo calls as well, you would have noticed the calls will only be triggered by scrolling for the magnified pixel (10 pixels instead of 1). Which still means that resetting the zoom level back to 1 after zooming has completed solved all problems. Also your problem IS the kvo calls because you are updating the position of the markers based on kvo events (thevmarkers are not even inside the scroll view). You can actually test this by zooming in heavily and scroll for less then 10 pixels and you will see no kvo events triggered untill you scroll enough pixels to cover an 'enlarged' pixel.

The markers are not inside the scroll view because they would be scaled while zooming. They are on an entirely different layer.

**Comment by incanus** [comment]
I am also resetting the zoom level back to 1 and also not drawing a bitmap, but rather using sublayers.

**Comment by specialunderwear** [comment]
Awesome, then in any case I'm positive the problem will be fixed. Not sure if having layers will change anything since cglayers are rendered to the pixel buffer as a bitmap in the end, at which point you can still do transformations on them.

So most likely you will be zooming and panning in the scrollview with the allready rendered layers. But yeah cglayer rendering is really fast.

Good luck!

**Comment by incanus** [comment]
Will keep at it. Thanks for the back-and-forth; will keep progress updated here.

**Comment by specialunderwear** [comment]
Hi how is it going? Any progress?

**Comment by cliffrowley** [comment]
Is there a workaround for this?  Looking forward to a solution...

**Comment by incanus** [comment]
No workaround yet.

**Comment by specialunderwear** [comment]
Are you working on this in a branch or fork? I'd like to help.

**Comment by incanus** [comment]
I'm not actively working on it right now, as I'm preparing the next release of the SDK. After that, I'm targeting this and tile performance together since both will be somewhat disruptive. 

If you'd like to work on it in the meantime, that's good. I can share some of the ideas I had, but I don't have code as of yet.

**Comment by specialunderwear** [comment]
Ok, I can work on it on the Alpstein fork. Please share your ideas. My email is in my github profile.

**Comment by incanus** [comment]
I'll just post them here. If you look at `RMLoadingTileView`, which is the gridded background behind loading map tiles, I'm thinking it could behave similarly. That view is a `UIScrollView` and it adjusts its content view after each pan and zoom to remain 3x the size of the super view and centered. 

I'm thinking the actual tiled view(s) could behave like this, too. It could track the supposed offset so that it knows which tiles to display. The only thing I'm worried about is pans and zooms over large areas, as I'm not sure it will still look convincing and may look like the reuse that it is. 

Does that make sense?

**Comment by tkohout** [comment]
I have found very dirty workaround for this. But I really needed it and it's working. The idea is to convert source from bigger zoom to smaller so the wiggle stops. Now it's working for only smaller areas and some small range of zoom. If anybody would be interested I can post the source codes.

**Comment by incanus** [comment]
I'd love to see it. I think that's basically what I'm going to try to do when I address this fully, effectively. The zoom will be legitimate but it won't zoom the scroll view as much as it does now. Feel free to [gist](https://gist.github.com) the code or paste inline or whatever.

**Comment by tkohout** [comment]
Ok here it is, it is for alpstein fork but it should work for mapbox too I guess. 

https://github.com/tkohout/route-me

There is sample Zoom Convert with some sample google map source.

Basic idea is to take base coordinate which will serve as zero point from where the map will stretch to every direction. Zoom steps is how much zoom levels you want to skip. Min zoom is there for converting base coordinate to proper base tile. Then every loaded tile is converted to proper tile from the bigger zoom.

All coordinates drawn on map (markers, boundaries, routes) have to be converted too.

I haven't tested it properly but it has basic idea.

**Comment by lemak5** [comment]
@incanus : do you know when this bug will be fixed?
Thank you !

**Comment by incanus** [comment]
No updates to report yet.

**Comment by jcouture** [comment]
Is there any known workaround or incoming fix for this issue?

**Comment by tkohout** [comment]
If you need to display only small area, for example one cite or state, it could help to move the city coordinates to the null coordinates where tile 0,0 is. The content offset would not get that high and the float issue would not appear. However, it requires change of map sources connectors code.

**Comment by incanus** [comment]
See also http://stackoverflow.com/questions/448285/how-do-i-reset-after-a-uiscrollview-zoom

**Comment by gthompson333** [comment]
Hi All.  It sounds like the best workaround to this issue is to dupe the scrollview into rendering tiles zoomed in, while it's zoomScale value is set to 1.  This way, we still get the finely granular contentOffset property changes, which helps alleviate the twitchy pin movement.  Is my nutshell solution description correct?

Any chance Mapbox might have this solution implemented soon?  Thanks.

**Comment by incanus** [comment]
I have been working on a similar approach using a scroll view to drive a GL-based view, where the scroll view never leaves the `0.5 - 2.0` zoom scale range. I may be able to integrate this soon, but the implications throughout `RMMapView` are pretty broad. Many things use the scroll view values directly. 

I've also spoken just yesterday with someone who fixed this internally a different way and am trying to get the solution contributed back.

**Comment by gthompson333** [comment]
Hey incanus.  Thanks much for the info.

It seems, from a high level perspective, what is happening is, the scrollview, by design, is adjusting the frequency of contentOffset events to be in proportion to the zoomLevel.  This actually makes sense.  At zoom level 1, a point of contentOffset would be different than at zoom level 10.  That is, at zoom level 1, contentOffset events happen for every point of scroll, but at zoom level 10, one would expect contentOffset events to happen every 10 points of scroll.

Because pin markers are often not scaled, along with the zoom scale, the "every 10 points" of scroll, when contentOffset events are generated, is very pronounced.  That is, twitchy pin marker movements.

I'm wondering if the ideal solution, albight not quickest, might be to eventually ditch the UIScrollview altogether, and implement your own scrolling image container.

I'm experimenting with some ideas.  I'll be sure to keep you informed.  Thanks much for your efforts in resolving this.

**Comment by incanus** [comment]
That's kind of what I was thinking -- use the scroll view for gestures and "bounciness", but maintain the world offset, bounds, and zoom independently of the scroll view's properties.

**Comment by abetlehem** [comment]
We had the same issue when overlaying layers with high-resolution tilesets, up to zoom level 22. When you zoomed, the tilesets would be very jumpy and stretched.
To fix: Make sure all the overlay tilesets have alpha of 1.0. Instantly smooth scrolling. Not sure why but it worked for us.

**Comment by cga-F4** [comment]
@incanus : now that v1.1.0 is released (we moved to this version for our project), do you plan to work on this bug ?

**Comment by florianbuerger** [comment]
Are there any updates for this issue? We are developing an applications for exhibitions and we are considering to use Mapbox & custom mbtiles for offline indoor maps. 

Sadly this issue is a deal breaker :disappointed:

**Comment by dpfaffenbauer** [comment]
Are there any updates for this issue?

**Comment by florianbuerger** [comment]
@dpfaffenbauer I think since they released a completely new renderer written in OpenGL they focus on that instead.

**Comment by incanus** [comment]
Yes, and the goal is to merge the APIs making a clean upgrade path from the SDK to GL.

**Comment by florianbuerger** [comment]
Nice. Looking forward to the future of the new renderer.

**Comment by dpfaffenbauer** [comment]
Sounds great :), looking forward to it

**Comment by mohpor** [comment]
Can't wait to use it!

**Comment by incanus** [comment]
I should point out that this renderer is available now: 

https://github.com/mapbox/mapbox-gl-cocoa

It's just not as full-featured yet as we're intending. Soon!

**Comment by specialunderwear** [comment]
Awesome go mapbox!

**Comment by incanus** [comment]
BTW any issue in this repo tagged with `v` is a nod to something that is more fundamental in the current renderer that can be solved by the new (vector) one (i.e., writing in OpenGL instead of using Core Graphics basing things on `UIScrollView` & `CATiledLayer`).

**Comment by incanus** [comment]
For some clarification on roadmap, I just added this: https://github.com/mapbox/mapbox-gl-cocoa/blob/master/FAQ.md

**Comment by mohpor** [comment]
Thanks @incanus for the FAQ, but it didn't answer one of my most important questions: What iOS versions will it support? The current repo says iOS 7+, will it be the same after release SDK 2.x?

**Comment by incanus** [comment]
@mohpor I've just updated the FAQ. GL will support iOS 7 & greater, which means that moving from the iOS SDK will leave behind iOS 5 and 6.

**Comment by mohpor** [comment]
Such a bad news :(

**Comment by incanus** [comment]
You can keep using the SDK for the older versions, but by the time GL is stable, iOS 8 will be out. There is already 90%+ upgrade to iOS 7 in less than a year after release, and iOS 8 only dropped the iPhone 4, so I expect similar numbers for upgrades this fall. While GLKit is available back to iOS 5, supporting iOS 7+ lets us focus on UI and UX testing for the modern platforms instead of a lot of code branching like in the iOS SDK to support the dual appearances.

**Comment by mohpor** [comment]
I see what you're saying, but as you sure know, Location Aware apps are fundamental applications that need to cover a larger than usual demographic...
I hope i can try and make it work under at least iOS 6.
Thanks for the effort and let me know if I could be of any help. I have tried and answered some issues here and on SO, to share some daily burden and make you free to work on 2.x..
Cheers

**Comment by incanus** [comment]
Yes, your help has been very much appreciate @mohpor and I am happy to work with you to help make GL unofficially available to earlier versions.

**Comment by mohpor** [comment]
Glad to help.

**Comment by pawel-sp** [comment]
Is there any working solution for shaking RMAnnotations problem ?

**Comment by incanus** [comment]
@pawel-sp No, not currently. This is a possible interim approach: 

https://github.com/mapbox/mapbox-ios-sdk/issues/72#issuecomment-14188749

**Comment by BalestraPatrick** [comment]
I'm having the same problem with my RMAnnotations and it's really annoying. Any fix or workaround?

**Comment by dcacenabes** [comment]
in iOS8 the problem is gone.

**Comment by incanus** [comment]
I have noticed some changes to `UIScrollView`, including better debug logging. I'll bet they increased the precision to deal with potential larger content views for larger and denser displays.

**Comment by dcacenabes** [comment]
I just found out that the problem is only gone in the simulator. In a iphone 5 with iOS8 the problem still exists.

**Comment by incanus** [comment]
Dang. 

To update everyone and summarize the above, this is a weakness of using a literal scroll view `contentSize` that maps to the actual pixel size of the whole planet once you get really far zoomed in. Each map zoom level is `(2 ^ z) * 256` pixels on an edge, so `z17` is `33,554,432` pixels, which runs into `float` precision problems. 

This is solved in [Mapbox GL](https://github.com/mapbox/mapbox-gl-cocoa/blob/master/FAQ.md) by rendering in an OpenGL context on screen instead of a content view with virtual pixel size. Migration path coming soon...

**Comment by tabsong** [comment]
hi: incanus
You said:“Each map zoom level is (2 ^ z) \* 256 pixels on an edge, so z17 is 33,554,432 pixels, which runs into float precision problems” , this conclusion is correct, but why we can't break through this precision problems. I encounter the same problem on our 2D Map SDK. using the same technology. UIScrollView+CATiledLayer. so our problem is the same. But now we have a Solution: 

This solution has a basic precondition: using double/long double to convert geographic coordinate system to UIScrollView coordinate system.

And Then:  We Just need to replenish the difference using translation transform. directly to reset center using "center" property is invalid.

```
// Get the coordinate of annotation.
CLLocationCoordinate2D coo = [annotationView.annotation coordinate];

// using long double to convert geographic coordinate system to UIScrollView coordinate system.
long double reX, reY;
[self convertLat:coo.latitude lon:coo.longitude to:&reX y:&reY];

reX += annotationView.centerOffset.x;
reY += annotationView.centerOffset.y;
```

   // Set center first
    annotationView.center = CGPointMake(reX, reY);

```
// Retrive the center.
CGPoint after = annotationView.center;

// Calculate the difference. We know the center will be cut to times of 2.
double dx = reX - after.x;
double dy = reY - after.y;

// Construct a translation transform.
CGAffineTransform translation = CGAffineTransformMakeTranslation(dx, dy);
```

   // replenish the difference using translation transform
    annotationView.transform = translation;

Yea: using the above code. The joggle is disappeared. 

Good Luck!

**Comment by incanus** [comment]
Yes @tabsong this is what is described above in https://github.com/mapbox/mapbox-ios-sdk/issues/72#issuecomment-29924109. However this is a pretty major refactoring and not just happening at one place in the map view code. I'd love to see this happen but we don't have the resources for it presently.

**Comment by tabsong** [comment]
@incanus Sounds like sadness, But i respect your choice. May be some days later. at lease there exists a non-elegant solution.

**Comment by leandersikma** [comment]
So we can conclude that this bug will stay unsolved. The solutions that are described earlier in this thread are too much work to make it reasonable and the Mapbox team is focussing on the MapboxGL migration. @incanus, you're saying that this bug doesn't occur in MapboxGL, right? 

Unfortunately, according to the documentation MapboxGL isn't production ready and misses some features such as offline maps and annotation clustering. So, it's either using the standard Mapbox with dancing markers or use MapboxGL and hope that things as offline maps are ready in the upcoming months. Am I right at all of this?

@florianbuerger What did you decide to do to built your indoor map app?

**Comment by incanus** [comment]
> @incanus, you're saying that this bug doesn't occur in MapboxGL, right?

Correct. This bug is inherent to using `UIScrollView` for maps and having precision problems with its API at high zoom levels because the content view is so large. Mapbox GL doesn't scroll, but actually redraws individual frames with OpenGL in response to zoom, pan, and rotation changes. 

> hope that things as offline maps are ready in the upcoming months

Yes. Plans for GL include offline maps and most likely clustering.

**Comment by leandersikma** [comment]
Okay, thanks. 

The FAQ on MapboxGL says "The library is open source right now, but an official, production-recommended release will come later in 2015.". Does this mean that there is a big possibility that offline map support will be implemented before Q3 2015? Or is that just too optimistic?

**Comment by incanus** [comment]
I think Q3 2015 is a pretty fair estimate. We are considering offline use and programmatic runtime style API in GL to be flagship features, but post-1.0, and are looking at a 1.0 by midyear at latest, I'd say.

**Comment by Svantulden** [comment]
I'm beginning to suspect this issue is because of 32/64bit architecture. If the problem is apparent on iPhone 4 and 5, but not on 5S or 6 regardless of iOS version, it could be that.

**Comment by braandl** [comment]
@Svantulden Yes, it is exactly that way. 
The CGFloat in the Core Graphics on 64Bit Systems (e.g. iPhone6) is internally a double value, but on older 32 Bit devices it is a float (due to performance reasons).
The Float is not precise enough to allow this kind of behavior that the ios-sdk uses, but the double value are...

**Comment by incanus** [comment]
Just a note that this problem can be circumvented with the rapidly-maturing [Mapbox Mobile](https://www.mapbox.com/mobile/) OpenGL-based renderer, which can also [do raster tiles](https://www.mapbox.com/mapbox-gl-ios/examples/raster-styles/).
